### PR TITLE
Report error diagnostic when trying to flattening a model with polymorphsim

### LINKED ISSUE #959: [TCGC] propertyFlatten not apply to property of polymorphic type
### Clear and concise description of the problem

Context: at present, `@flattenProperty` had to be applied to template property. e.g. https://github.com/Azure/azure-rest-api-specs/blob/1418642cf8a57954c781c26c759fd4aef39a1287/specification/vmware/Microsoft.AVS/client.tsp#L19

For a property of polymorphic type, it is not possible to flatten the model, as polymorphic is runtime behavior.
https://github.com/Azure/azure-rest-api-specs/blob/1418642cf8a57954c781c26c759fd4aef39a1287/specification/vmware/Microsoft.AVS/models.tsp#L1715-L1744

Therefore, either the decorator, or the TCGC, may want to not set `SdkBodyModelPropertyType.flatten=true`.

Alternatively, this can be done in language emitter part (Java does this), as e.g. one would still encounter it from Swagger. However, the flatten may already be done by m4.
https://github.com/Azure/azure-rest-api-specs/blob/32f566eb38af9593cfc1ed183362471c50d67fb2/specification/vmware/resource-manager/Microsoft.AVS/stable/2023-09-01/vmware.json#L6514-L6516

### Checklist

- [X] Follow our [Code of Conduct](https://github.com/azure/typespec-azure/blob/main/CODE_OF_CONDUCT.md)
- [X] Check that this issue is about the Azure libraries for typespec. For feature request in the typespec language or core libraries file it in the [TypeSpec repo](https://github.com/Microsoft/TypeSpec/issues/new/choose)
- [X] Read the [docs](https://azure.github.io/typespec-azure/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by archerzz**
resolve #959

### CODE REVIEW

File: .chronus/changes/tcgc-forbid-polymorphism-flatten-2024-6-23-22-12-38.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+changeKind: feature
+packages:
+  - "@azure-tools/typespec-client-generator-core"
+---
+
+throw error when trying to flattening a model with polymorphsim
```
Comment (timotheeguerin) [inline .chronus/changes/tcgc-forbid-polymorphism-flatten-2024-6-23-22-12-38.md:7]:
```suggestion
Report error diagnostic when trying to flattening a model with polymorphism
```

do not call reporting diagnostic throw as this is something you should absolutely not do in TypeSpec libraries.

File: packages/typespec-client-generator-core/test/decorators.test.ts [hunk 2]
```diff
@@ -2545,6 +2545,34 @@ describe("typespec-client-generator-core: decorators", () => {
         code: "decorator-wrong-target",
       });
     });
+
+    it("throws error when used on a polymorphism type", async () => {
+      const diagnostics = await runner.diagnose(`
+        @service({})
```
Comment (timotheeguerin) [inline packages/typespec-client-generator-core/test/decorators.test.ts:2551]:
```suggestion
        @service
```

File: packages/typespec-client-generator-core/test/decorators.test.ts [hunk 3]
```diff
@@ -2545,6 +2545,34 @@ describe("typespec-client-generator-core: decorators", () => {
...
+
+          @test
+          @discriminator("kind")
+          model Model2{
+            kind: string;
+          }
+
+          @test
+          @route("/func1")
+          op func1(@body body: Model1): void;
```
Comment (timotheeguerin) [inline packages/typespec-client-generator-core/test/decorators.test.ts:2568]: is the route relevant to this test, can it be removed

File: packages/typespec-client-generator-core/src/lib.ts [hunk 4]
```diff
@@ -166,6 +166,12 @@ export const $lib = createTypeSpecLibrary({
         nonDecorator: paramMessage`Client name: "${"name"}" is defined somewhere causing nameing conflicts in language scope: "${"scope"}"`,
       },
     },
+    "flatten-polymorphism": {
+      severity: "error",
+      messages: {
+        default: `Flattening polymorphic types is not supported.`,
```
Comment (weidongxu-microsoft) [inline packages/typespec-client-generator-core/src/lib.ts:172]: nit, "Cannot flatten property of polymorphic type"?

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by archerzz** [comment]
I changed the merge target branch from `main` to `release/july-2024`, per suggestion.

**Comment by iscai-msft** [comment]
@timotheeguerin should this be an error, or should it be a linter warning? It is a bad thing to generate, but I'm not sure we should completely blow up generation in this case

**Comment by timotheeguerin** [comment]
I think it's fine to be an error if it's run in the validator as it would count as a validator of the for @flatten decorator. If run in the preprocessor then no. And people will not see it in the ide too

**Review by timotheeguerin — APPROVED** [review]

**Pull request merged**